File: Exercise_2.py
# Problem2
# Longest Substring Without Repeating Characters(https://leetcode.com/problems/longest-substring-without-repeating-characters/)
# // Time Complexity :O(2n)
# // Space Complexity :O(n)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No


# // Your code here along with comments explaining your approach
# Hashset Approach
class Solution:
    def lengthOfLongestSubstring(self, s):
        # creating a hashset of all uniques
        uniques = set()
        slow = 0
        length = 0
        # iterating over the string
        for i in range(len(s)):
            # till the char is in the set
            while s[i] in uniques:
                # we remove it
                uniques.remove(s[i])
                # update slow ptr
                slow += 1
            uniques.add(s[i])
        length = max(length, len(uniques))
        return length


sol = Solution()
print(sol.lengthOfLongestSubstring("abcabcbb"))

File: Exercise_2.cpp
/* 
    Time Complexity                              :  O(N) where N is the size of the string
    Space Complexity                             :  O(N) used in the hashmap
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // return usingHashMap(s);
        return usingArrays(s);
    }
    
    int usingHashMap(string s) {
        int l=0, r=0;
        unordered_map<char, int> mp;
        int mx = INT_MIN;
        while(r < s.size()) {
            char ch = s[r];
            mp[ch]++;
            while(mp[ch] > 1) {
                mp[s[l]]--;
                l++;
            }
            mx = max(mx, r-l+1);
            r++;
        }
        return mx == INT_MIN ? 0 : mx;
    }
    
    
    int usingArrays(string s) {
        vector<int> freq(256,0);
        int l=0,r=0,mx=0;
        while(r < s.size()) {
            freq[s[r]]++;
            while(freq[s[r]] > 1) {
                freq[s[l]]--;
                l++;
            }
            mx = max(mx, r-l+1);
            r++;
        }
        
        return mx;
    }
};